crates/execution/batch-validator/src/validator.rs

```diff
@@ -108,15 +110,66 @@ where
         Ok(())
     }
 
+    /// Validate the size of transactions (in bytes).
+    fn validate_batch_size_bytes(
+        &self,
+        transactions: &[Vec<u8>],
+        timestamp: u64,
+    ) -> BatchValidationResult<()> {
+        // calculate size (in bytes) of included transactions
+        let total_bytes = transactions
+            .iter()
+            .map(|tx| tx.len())
+            .reduce(|total, size| total + size)
+            .ok_or(BatchValidationError::EmptyBatch)?;
+        let max_tx_bytes = max_batch_size(timestamp);
+
+        // allow txs that equal max tx bytes
+        if total_bytes > max_tx_bytes {
+            return Err(BatchValidationError::HeaderTransactionBytesExceedsMax(total_bytes));
+        }
+
+        Ok(())
+    }
+
+    /// Decode transactions to ensure encode/decode is valid.
+    ///
+    /// The decoded transactions are then used to validate max batch gas.
+    #[inline]
+    fn decode_transactions(
+        &self,
+        transactions: &Vec<Vec<u8>>,
+        digest: BlockHash,
+    ) -> BatchValidationResult<Vec<TransactionSigned>> {
+        if transactions.len() < *PARALLEL_SENDER_RECOVERY_THRESHOLD {
+            transactions
+                .iter()
+                .map(|tx| Self::recover_and_validate(tx, digest))
+                .collect::<BatchValidationResult<Vec<_>>>()
+        } else {
+            transactions
```

Comment:
I think you are using rayon ok here but just want to point out that mixing rayon (a thread pool) and Tokyo (also a thread pool and async runtime) can lead to nasty interactions and bugs.  Again, I can't see how this simple use would be a problem but it might make it easier to foot gun ourselves later if we start using rayon more.  Just something worth considering.